### ISSUE TITLE
legion/realm: (optionally?) avoid the use of `mutable` in device code

### ISSUE THREAD

**Issue by streichler** [StanfordLegion/legion#1034]
The use of `mutable` fields in device code can be problematic if something (e.g. Kokkos in this case, but
I don't think what they're doing is unreasonable or unique) takes advantage of the `const`-ness of a
parameter and puts it in actually-read-only memory on the device.  In a perfect world we would avoid
use of `mutable` entirely in application code (it _is_ kind of an abomination), but if the decision comes with
tradeoffs, we at least need a compile-time option we can set that does the best it can without using
`mutable` fields.

The known failing case is the non-dense bounds check warning in Legion's device-side accessor code, but
I'm assigning this to myself as well because I need to scrub the Realm accessor stuff as well.

**Comment by lightsighter** [comment]
So at least in the `AffineAccessor` cases I think I can trivially remove the `mutable` annotations. However, for the `GenericAccessor` cases, most of the important methods are not marked `const` so I can't remove the `mutable` annotations. Why is it the case that things like `read` and `write` on `GenericAccessor` are not marked `const`? Nothing should be modifying the accessor itself.
https://gitlab.com/StanfordLegion/legion/-/blob/master/runtime/realm/inst_layout.h#L438-439
The invariant I would like to maintain on the Legion side is that all the `read/write/etc` methods have the same type signatures regardless of the Realm accessor specialization.

**Comment by streichler** [comment]
Those methods are not const because of the caching that goes on here: https://gitlab.com/StanfordLegion/legion/-/blob/master/runtime/realm/inst_layout.h#L450-455

Const versions of those methods would have to do piece lookups on every single read/write call, which would make the poor
performance of a GenericAccessor even more terrible.  What's the benefit in making the Realm accessor inside a Legion
accessor more const than the wrapper?

**Comment by lightsighter** [comment]
Well, I'm pretty sure these `const` annotations were there from a very old version of Legion where all the Realm accessors were also `const` (e.g. from a time before there were fields or layouts or anything like that). Changing it now would break backwards compatibility of the interface. There are lots of codes out there that do `const FieldAccessor<...>`. Using `mutable` was the way for me to adapt to the new Realm accessors and not break backwards compatibility for Legion users.

**Comment by lightsighter** [comment]
If it were me, I would dangle the cache as a separate object pointed to by the accessor. That way the accessor can have all its methods be marked `const` but then there can be non-`const` changes to the cache.

**Comment by lightsighter** [comment]
Another though: if you decouple the cache from the accessor, you can copy the accessor without copying the cache, say for example copying it from the CPU down to the GPU. The cache can then be built back up wherever it is being used without needing to copy its contents around everywhere. That way accessors stay lightweight and easy to move around.

**Comment by streichler** [comment]
>Changing it now would break backwards compatibility of the interface. There are lots of codes out there that do const FieldAccessor<...>. Using mutable was the way for me to adapt to the new Realm accessors and not break backwards compatibility for Legion users.

Yeah, I figured we might not be able to change the default, but we do need a mode that doesn't use mutable, and I think it's ok if that has slightly different usage rules.  At the very least, is it possible to remove this usage of mutable on `gpu_warning`?  That's the one that's currently breaking things.

**Comment by lightsighter** [comment]
The `gpu_warning` is there to make sure you don't get swamped with warning messages in the case that you aren't actually going to die. It's a bit of state that says whether we issued the warning or not. I don't think I can make it non-mutable without either getting rid of it or changing the top-level `const` signature of the `read/write/operator[]` methods. So yes, we could get rid of it at some cost of niceness to the user.

Aside: you should only be seeing this bit of state if you build with bounds checks on. If you think bounds checks should be off then something has gone wrong.

What exactly is the error that the compiler is reporting?

**Comment by streichler** [comment]
Yes, the `gpu_warning` mutable-ness is only problematic when bounds checks are enabled, but it's not a compile error - it's a fatal error at runtime because Kokkos used the 'const'-ness of the overall accessor as justification for moving the whole thing to CUDA constant memory, which does not support writes at all.  The feedback from the Kokkos folks is that Kokkos doesn't support `mutable`, so we need a way to not use mutable even if it results in a minor loss of functionality (e.g. preventing us from issuing this warning about imprecise bounds checks).

**Comment by lightsighter** [comment]
I've pushed a change master that removes the superfluous `mutable` annotations on `Realm::AffineAccessor`s and `Realm::MultiAffineAccessor`s. I also deleted the `gpu_warning` flag for now, so that `mutable` annotation is gone as well. There are still `mutable` annotations though on `Realm::GenericAccessor`s because those are required for the backwards compatibility of the interface. We can't use `Realm::GenericAccessor`s in device code anyway right now, so should not be an issue in the short term. We should revisit this when `Realm::GenericAccessor`s can be used in device code.
